### PR TITLE
Datasets

### PR DESCRIPTION
**Pull request by samwang141224**
add the codes for generating datasets

### CODE REVIEW

File: desidlas/datasets/datasetting.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    REST_RANGE: list
+    kernel: int, optional
+
+    Returns
+    -------
+
+    """
+    lam, lam_rest, ix_dla_range = get_lam_data(sightline.loglam, sightline.z_qso, REST_RANGE)
+    kernelrangepx = int(kernel/2) #200
+    #samplerangepx = int(kernel*pos_sample_kernel_percent/2) #60
```
Comment (profxj) [inline desidlas/datasets/datasetting.py]: remove this code?

File: desidlas/datasets/datasetting.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    #padding the sightline:
+    flux_padded,lam_padded,pixel_num_left=pad_sightline(sightline,lam,lam_rest,ix_dla_range,kernelrangepx,v=best_v['b'])
+     
+    #ix_dlas = [(np.abs(lam[ix_dla_range]-dla.central_wavelength).argmin()) for dla in sightline.dlas]
```
Comment (profxj) [inline desidlas/datasets/datasetting.py]: remove this commented code?

File: desidlas/datasets/datasetting.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    #coldensity_dlas = [dla.col_density for dla in sightline.dlas]       # column densities matching ix_dlas
+
+    # FLUXES - Produce a 1748x400 matrix of flux values
+    #fluxes_matrix = np.vstack(map(lambda x:x[0][x[1]-kernelrangepx:x[1]+kernelrangepx],zip(itertools.repeat(sightline.flux), np.nonzero(ix_dla_range)[0])))
+    fluxes_matrix = np.vstack(map(lambda x:x[0][x[1]-kernelrangepx:x[1]+kernelrangepx],zip(itertools.repeat(flux_padded), np.nonzero(ix_dla_range)[0]+pixel_num_left)))
+    lam_matrix = np.vstack(map(lambda x:x[0][x[1]-kernelrangepx:x[1]+kernelrangepx],zip(itertools.repeat(lam_padded), np.nonzero(ix_dla_range)[0]+pixel_num_left)))
+    #using cut will lose side information,so we use padding instead of cutting 
+    #fluxes_matrix = np.vstack(map(lambda x:x[0][x[1]-kernelrangepx:x[1]+kernelrangepx],zip(itertools.repeat(sightline.flux), np.nonzero(ix_dla_range)[0][cut])))
```
Comment (profxj) [inline desidlas/datasets/datasetting.py]: remove?

File: desidlas/datasets/datasetting.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    #lam_matrix = np.vstack(map(lambda x:x[0][x[1]-kernelrangepx:x[1]+kernelrangepx],zip(itertools.repeat(lam), np.nonzero(ix_dla_range)[0][cut])))
+    #the wavelength and flux array we input:
+    input_lam=lam_padded[np.nonzero(ix_dla_range)[0]+pixel_num_left]
+    input_flux=flux_padded[np.nonzero(ix_dla_range)[0]+pixel_num_left]
+    # Return
+    return fluxes_matrix, sightline.classification, sightline.offsets, sightline.column_density,lam_matrix,input_lam,input_flux
+    #return fluxes_matrix, sightline.classification, sightline.offsets, sightline.column_density
```
Comment (profxj) [inline desidlas/datasets/datasetting.py]: remove

File: desidlas/datasets/datasetting.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    classification: np.ndarray
+        Array of classification values.  1=DLA; 0=Not; -1=not analyzed
+
+    Returns
+    -------
+    idx: np.ndarray
+        positive + negative indices
+
+    """
+    #classification = data[1]
```
Comment (profxj) [inline desidlas/datasets/datasetting.py]: remove this and old, commented out code

File: desidlas/datasets/datasetting.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    lam, lam_rest, ix_dla_range = get_lam_data(sightline.loglam, sightline.z_qso)
+    kernelrangepx = int(kernel/2)
+    #cut=((np.nonzero(ix_dla_range)[0])>=kernelrangepx)&((np.nonzero(ix_dla_range)[0])<=(len(lam)-kernelrangepx-1))
+    #newclassification=sightline.classification[cut]
+    num_pos = np.sum(sightline.classification==1, dtype=np.float64)
+    num_neg = np.sum(sightline.classification==0, dtype=np.float64)
+    n_samples = int(min(num_pos, num_neg))
+
+    r = np.random.permutation(len(sightline.classification))
```
Comment (profxj) [inline desidlas/datasets/datasetting.py]: Add some in-line comments

File: desidlas/datasets/datasetting.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+    pos_ixs = r[sightline.classification[r]==1][0:n_samples]
+    neg_ixs = r[sightline.classification[r]==0][0:n_samples]
+    # num_total = data[0].shape[0]
+    # ratio_neg = num_pos / num_neg
+
+    # pos_mask = classification == 1      # Take all positive samples
+
+    # neg_ixs_by_ratio = np.linspace(1,num_total-1,round(ratio_neg*num_total), dtype=np.int32) # get all samples by ratio
```
Comment (profxj) [inline desidlas/datasets/datasetting.py]: clean up all this commented code

File: desidlas/datasets/get_dataset.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+
+def make_datasets(sightlines,validate=True):
+    """
+    Generate training set or validation set for DESI.
+    
+    Parameters:
+    -----------------------------------------------
+    sightlines: list of 'dla_cnn.data_model.Sightline' object, the sightlines should be preprocessed.
+    validate: bool
```
Comment (profxj) [inline desidlas/datasets/get_dataset.py]: add `,optional`  and describe what this does

File: desidlas/datasets/get_dataset.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    
+    Returns
+    -----------------------------------------------
+    dataset:dict, the training set contains flux and 3 labels, the validation set contains flux, lam, 3 labels and DLAs' data.
+    
+    """
+    dataset={}
+    for sightline in sightlines:
+        wavelength_dlas=[dla.central_wavelength for dla in sightline.dlas]
```
Comment (profxj) [inline desidlas/datasets/get_dataset.py:25]: what are the `wavelength_dlas` and `coldensity_dlas` lists for?
can they be removed?

File: desidlas/datasets/get_dataset.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        if validate:
+            flux=np.vstack([data_split[0]])
+            labels_classifier=np.hstack([data_split[1]])
+            labels_offset=np.hstack([data_split[2]])
+            col_density=np.hstack([data_split[3]])
+            lam=np.vstack([data_split[4]])
+            dataset[sightline.id]={'FLUX':flux,'lam':lam,'labels_classifier':  labels_classifier, 'labels_offset':labels_offset , 'col_density': col_density,'wavelength_dlas':wavelength_dlas,'coldensity_dlas':coldensity_dlas} 
+        else:
+            sample_masks=select_samples_50p_pos_neg(sightline)
+            if sample_masks !=[]:
```
Comment (profxj) [inline desidlas/datasets/get_dataset.py]: how about `if len(sample_masks) > 0`

File: desidlas/datasets/get_dataset.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+            dataset[sightline.id]={'FLUX':flux,'labels_classifier':labels_classifier,'labels_offset':labels_offset,'col_density': col_density}
+    return dataset
+        
+def smooth_flux(flux):
+    """
+    Smooth flux using median filter.
+    
+    Parameters:
+    -----------------------------------------------
+    flux: np.ndarry flux for every kernel
```
Comment (profxj) [inline desidlas/datasets/get_dataset.py:60]: is this a 2D array?  If so, describe the dimensions

File: desidlas/datasets/get_dataset.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    
+    Return:
+    -----------------------------------------------
+    flux_matrix:list, 4-dimension flux data
+    
+    """
+    flux_matrix=[]
+    for sample in flux:#sample是片段flux
+                smooth3=signal.medfilt(sample,3)
```
Comment (profxj) [inline desidlas/datasets/get_dataset.py]: don't double indent

File: desidlas/datasets/get_dataset.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+    -----------------------------------------------
+    flux: np.ndarry flux for every kernel
+    
+    Return:
+    -----------------------------------------------
+    flux_matrix:list, 4-dimension flux data
+    
+    """
+    flux_matrix=[]
+    for sample in flux:#sample是片段flux
```
Comment (profxj) [inline desidlas/datasets/get_dataset.py]: I'm not sure what `flux` is, but you are looping over it
and creating `smooth3`, etc. that aren't used until 
after you are done looping.

Something looks wrong here..

File: desidlas/datasets/get_dataset.py [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+                smooth3=signal.medfilt(sample,3)
+                smooth7=signal.medfilt(sample,7)
+                smooth15=signal.medfilt(sample,15)
+                flux_matrix.append(np.array([sample,smooth3,smooth7,smooth15]))
+    return flux_matrix
+    
+def make_smoothdatasets(sightlines,validate=True):
```
Comment (profxj) [inline desidlas/datasets/get_dataset.py:76]: does this need to be a separate method?

Would be best if you could modify `make_datasets` to take
a `smooth` boolean and avoid duplicating code.

File: desidlas/datasets/get_sightlines.py [hunk 15]
```diff
@@ -0,0 +1,45 @@
+from desidlas.datasets.preprocess import estimate_s2n,normalize,rebin
+from desidlas.datasets.DesiMock import DesiMock
+from desidlas.dla_cnn.defs import best_v
+import numpy as np
+import os
+from os.path import join
+from pkg_resources import resource_filename
+
+
+datafile_path = os.path.join(resource_filename('desidlas', 'tests'), 'datafile')
```
Comment (profxj) [inline desidlas/datasets/get_sightlines.py]: Why are these lines here?

File: desidlas/datasets/get_sightlines.py [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+#load spectra file for testing
+spectra= os.path.join(datafile_path, 'spectra', 'spectra-16-1375.fits')
+truth=os.path.join(datafile_path, 'spectra', 'truth-16-1375.fits')
+zbest=os.path.join(datafile_path, 'spectra', 'zbest-16-1375.fits')
+
+
+def get_sightlines(spectra,truth,zbest):
+    """
+    Insert DLAs manually into sightlines without DLAs or only choose sightlines with DLAs
+
```
Comment (profxj) [inline desidlas/datasets/get_sightlines.py:6]: Add `Parameters` and explain them.

File: desidlas/tests/test_dataset.py [hunk 17]
```diff
@@ -0,0 +1,48 @@
+#load basic module
+import numpy as np
+import scipy.signal as signal
+from desidlas.datasets.datasetting import split_sightline_into_samples,select_samples_50p_pos_neg,pad_sightline
+from desidlas.datasets.preprocess import label_sightline
+from desidlas.dla_cnn.spectra_utils import get_lam_data
+
+#load the data for the test
+#this file is the output of get_sightlines.py, modules are tested in test_preprocess
+sightline_path='desidlas/tests/datafile/sightlines-16-1375.npy'
```
Comment (profxj) [inline desidlas/tests/test_dataset.py]: you can't hard-code the path like this.  Best to use
`resource_filename` for the root path of the package.

File: desidlas/tests/test_dataset.py [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+
+sightlines=np.load(sightline_path,allow_pickle = True,encoding='latin1')
+
+##load modules in get_datasets.py
+from desidlas.datasets.get_dataset import make_datasets
+from desidlas.datasets.get_dataset import make_smoothdatasets
+
+#test the flux,lam and the labels in the dataset
+#labels:FLUX,lam,labels_classifier,labels_offset,col_density,wavelength_dlas,coldensity_dlas
+def test_labels(spec):
```
Comment (profxj) [inline desidlas/tests/test_dataset.py]: rename this method to something like `check_labels`
otherwise it will confuse the CI system.

**Review by profxj — CHANGES_REQUESTED** [review]
Very good.  A number of comments from me
but only one or two that are significant.

File: desidlas/datasets/datasetting.py (hunk 1 above)
Comment (samwang141224) [reply desidlas/datasets/datasetting.py]: done

File: desidlas/datasets/datasetting.py (hunk 2 above)
Comment (samwang141224) [reply desidlas/datasets/datasetting.py]: done

File: desidlas/datasets/datasetting.py (hunk 3 above)
Comment (samwang141224) [reply desidlas/datasets/datasetting.py]: done

File: desidlas/datasets/datasetting.py (hunk 4 above)
Comment (samwang141224) [reply desidlas/datasets/datasetting.py]: done

File: desidlas/datasets/datasetting.py (hunk 5 above)
Comment (samwang141224) [reply desidlas/datasets/datasetting.py]: done

File: desidlas/datasets/datasetting.py (hunk 6 above)
Comment (samwang141224) [reply desidlas/datasets/datasetting.py]: done

File: desidlas/datasets/datasetting.py (hunk 7 above)
Comment (samwang141224) [reply desidlas/datasets/datasetting.py]: done

File: desidlas/datasets/get_dataset.py (hunk 8 above)
Comment (samwang141224) [reply desidlas/datasets/get_dataset.py]: done

File: desidlas/datasets/get_dataset.py (hunk 9 above)
Comment (samwang141224) [reply desidlas/datasets/get_dataset.py:25]: the wavelength_dlas and coldensity_dlas are the center wavelength and column density for the DLAs, we store them in the dataset to make it convenient to  validate

File: desidlas/datasets/get_dataset.py (hunk 10 above)
Comment (samwang141224) [reply desidlas/datasets/get_dataset.py]: thanks for the suggestion! I have updated the code here

File: desidlas/datasets/get_dataset.py (hunk 11 above)
Comment (samwang141224) [reply desidlas/datasets/get_dataset.py:60]: no, this flux is 1D array, the  flux_matrix is a 2D array, I have clarified this in the comments

File: desidlas/datasets/get_dataset.py (hunk 12 above)
Comment (samwang141224) [reply desidlas/datasets/get_dataset.py]: sorry, I have corrected this and removed the useless comments

File: desidlas/datasets/get_dataset.py (hunk 13 above)
Comment (samwang141224) [reply desidlas/datasets/get_dataset.py]: flux is the input for this module, every sample in flux is a 1D array for every window(flux array with length of 400 or 600), we loop every sample in flux and make the smoothing, then store the 2D array flux in flux_matrix

File: desidlas/datasets/get_sightlines.py (hunk 15 above)
Comment (samwang141224) [reply desidlas/datasets/get_sightlines.py]: sorry for the confusion, these codes are already tested in test_preprocess.py. I just want to make these code in a py file because this is an important step in making the dataset(load fits file and store them in a npy file ). For the users who want use our code, they will just need to 'from get_sightlines.npy import get_sightlines'. Then the npy file can be generated when they input the file path of spectra

File: desidlas/datasets/get_dataset.py (hunk 14 above)
Comment (samwang141224) [reply desidlas/datasets/get_dataset.py:76]: thanks for the suggestion, I have added a new boolean parameter smooth for this

File: desidlas/tests/test_dataset.py (hunk 18 above)
Comment (samwang141224) [reply desidlas/tests/test_dataset.py]: done, thanks!

File: desidlas/datasets/get_sightlines.py (hunk 16 above)
Comment (samwang141224) [reply desidlas/datasets/get_sightlines.py:6]: done

File: desidlas/tests/test_dataset.py (hunk 17 above)
Comment (samwang141224) [reply desidlas/tests/test_dataset.py]: thanks, I have corrected this

File: desidlas/datasets/get_dataset.py (hunk 9 above)
Comment (profxj) [reply desidlas/datasets/get_dataset.py:25]: got it

File: desidlas/datasets/get_sightlines.py (hunk 15 above)
Comment (profxj) [reply desidlas/datasets/get_sightlines.py]: sorry, I am still confused.  

At the very least, please put these lines into a new `def`

**Comment by profxj** [comment]
Almost there @samwang141224 

Just address that one comment in `get_sightlines.py` and I'll merge.

Thanks!

File: desidlas/datasets/get_sightlines.py (hunk 15 above)
Comment (samwang141224) [reply desidlas/datasets/get_sightlines.py]: excuse me, do you mean that I should make lines 7-14 in another py file but keep other lines still in get_sightline.py?

Comment (profxj) [reply desidlas/datasets/get_sightlines.py]: What I mean is that those lines shouldn't be there at the start of a module like this.

They will be run every time the module is imported which shouldn't be required
and is prone to cause a crash (if those files were moved, renamed, etc.)

Comment (samwang141224) [reply desidlas/datasets/get_sightlines.py]: Sorry I misunderstood you before. What I want to do with this get_sightlines.py file is for the people who want to use our codes, they will only need to run this file for the preprocessing. This may be more convenient to use because they don not need to check other py files for preprocessing. Maybe another way to achieve this goal is to remove this file and add a notebook. In the notebook we list every step for preprocessing just like in the test_preprocess.py . If the second way is better then I think I can just replace this get_sightlines.py with a notebook.

Comment (samwang141224) [reply desidlas/datasets/get_sightlines.py]: I make these lines here because I try to include every step for preprocessing in this py file. Then this one file is enough for the people who just want to run this code. But now I realized maybe a notebook is more proper for this? If so I can remove this file to avoid confusion and crash.

Comment (profxj) [reply desidlas/datasets/get_sightlines.py]: A Notebook sounds fine.  Thanks!

Comment (samwang141224) [reply desidlas/datasets/get_sightlines.py]: OK, I will remove this file and upload a notebook

Comment (samwang141224) [reply desidlas/datasets/get_sightlines.py]: I have edited get_sightlines.py and uploaded a notebook at desidlas/notebook to describe how to run the preprocessing codes

**Comment by profxj** [comment]
Remove the spaces in the name of your Notebook (can be problematic)
and I'll merge.  Thanks!

**Comment by samwang141224** [comment]
> Remove the spaces in the name of your Notebook (can be problematic)
> and I'll merge. Thanks!

Thanks for your remind. I have already renamed it.

**Review by profxj — APPROVED** [review]
